### PR TITLE
Fibonacci: Use Perl package

### LINKED ISSUE #3068: Fibonacci: Use perl package for generating sequence
Instead of rolling our own, we could use a package like [Math::NumSeq::Fibonacci](https://metacpan.org/pod/Math::NumSeq::Fibonacci) to generate the Fibonacci numbers.

------
IA Page:  http://duck.co/ia/view/fibonacci 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @koosha--

### PR DESCRIPTION
**Pull request by shellhead**
##### What does your Pull Request do (check all that apply)?
- [x] Improvement
    - [x] Enhancement: **`Fibonacci: Used the Perl Math::SeqNum::Fibonacci package to calculate Fibonacci numbers.`**

###### Description of changes
Removed logic for calculating Fibonacci numbers and replaced with the Perl package.

While working on this enhancement, I found that it did not trigger if uppercase characters were used, ie. "What is the 3rd fibonacci number". Also, returned early on the "nd" ordinal suffix, ie. "what is the 2nd fibonacci number". Resolved both of these issues. Also increased the limit from 1470 to 25000.

###### Which issues (if any) does this fix?
Fixes #3068 (which was more of an enhancement).

###### People to notify (@mention interested parties)
@mintsoft @GuiltyDolphin @edgesince84 

---

Instant Answer Page: https://duck.co/ia/view/fibonacci

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @koosha--

### CODE REVIEW

**Comment by daxtheduck** [comment]
@shellhead Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by edgesince84** [comment]
Hey @shellhead, thanks so much!  This is your first PR for DuckDuckHack, right? 🎈 🎉 🎈

**Comment by shellhead** [comment]
Thanks @edgesince84! The most recent commits also address issue #3070.

File: lib/DDG/Goodie/Fibonacci.pm [hunk 1]
```diff
@@ -5,37 +5,60 @@ use strict;
...
         }
     };
+}
+
+handle remainder_lc => sub {
+    my $ith_limit = 25000;      # limit for nth fibonacci numbers 
+    my $pred_limit = 10**22;    # limit for if n is a fibonacci number
+    my $fib_seq = Math::NumSeq::Fibonacci->new;
+
+    # trim whitespace around remainder
```
Comment (mintsoft) [inline lib/DDG/Goodie/Fibonacci.pm]: We have Text::Trim deployed if you'd prefer

File: t/Fibonacci.t [hunk 2]
```diff
@@ -23,13 +23,33 @@ sub build_test
     });
 }
 
+sub build_pred_test
```
Comment (mintsoft) [inline t/Fibonacci.t]: Can we refactor these into the same sub?

File: lib/DDG/Goodie/Fibonacci.pm [hunk 3]
```diff
@@ -5,37 +5,60 @@ use strict;
...
+    }
+    # check "is n a fibonacci number"
+    elsif (/^is (?<which>\d+) (?:a|in the)? ?(?:number|sequence)?\??$/ && $1 <= $pred_limit)
+    {
+        my $n = $+{'which'};
+        my $val = $fib_seq->pred($n);
+        my $is_fib = $val ? "is" : "is not";
+        my $text_answer ="$n $is_fib a Fibonacci number.";
+
+	return answer($text_answer, $val ? "true":"false", "$n $is_fib a Fibonacci number");
```
Comment (mintsoft) [inline lib/DDG/Goodie/Fibonacci.pm]: Isn't this last thing `$text_answer` ?

File: lib/DDG/Goodie/Fibonacci.pm (hunk 1 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Fibonacci.pm]: We shouldn't need to do this (`remainder_lc` should trim whitespace).

Comment (mintsoft) [inline lib/DDG/Goodie/Fibonacci.pm]: Good point!

File: lib/DDG/Goodie/Fibonacci.pm [hunk 4]
```diff
@@ -5,37 +5,60 @@ use strict;
...
+    }
+    # check "is n a fibonacci number"
+    elsif (/^is (?<which>\d+) (?:a|in the)? ?(?:number|sequence)?\??$/ && $1 <= $pred_limit)
+    {
+        my $n = $+{'which'};
+        my $val = $fib_seq->pred($n);
+        my $is_fib = $val ? "is" : "is not";
+        my $text_answer ="$n $is_fib a Fibonacci number.";
+
+	return answer($text_answer, $val ? "true":"false", $text_answer);
```
Comment (moollaza) [inline lib/DDG/Goodie/Fibonacci.pm]: I think a more casual "Yes" or "No" would be more appropriate, given that it's a question being asked?

File: lib/DDG/Goodie/Fibonacci.pm [hunk 5]
```diff
@@ -5,37 +5,60 @@ use strict;
...
+
+	return answer($text_answer, $val, "$n$suf Fibonacci number");
+    }
+    # check "is n a fibonacci number"
+    elsif (/^is (?<which>\d+) (?:a|in the)? ?(?:number|sequence)?\??$/ && $1 <= $pred_limit)
+    {
+        my $n = $+{'which'};
+        my $val = $fib_seq->pred($n);
+        my $is_fib = $val ? "is" : "is not";
+        my $text_answer ="$n $is_fib a Fibonacci number.";
```
Comment (moollaza) [inline lib/DDG/Goodie/Fibonacci.pm]: Let's drop the period when using this as a `title` -- we don't typically use periods for titles and subtitles

File: lib/DDG/Goodie/Fibonacci.pm (hunk 4 above)
Comment (shellhead) [inline lib/DDG/Goodie/Fibonacci.pm]: I thought responding with "false" made it sound like Dwight Schrute was responding. :)

**Comment by moollaza** [comment]
@shellhead thanks for the updates! I'll get this on Beta asap

**Comment by shellhead** [comment]
@moollaza of course! Let me know if I can be any more help!

File: t/Fibonacci.t [hunk 6]
```diff
@@ -9,27 +9,35 @@ use DDG::Test::Goodie;
...
+    'fib 7' => build_test("The 7th fibonacci number is 13 (assuming f(0) = 0)", 13, "7th Fibonacci number"),
+    'fibonacci 33' => build_test("The 33rd fibonacci number is 3524578 (assuming f(0) = 0)", 3524578, '33rd Fibonacci number'),
+    q`what's the 6th fibonacci number?` => build_test("The 6th fibonacci number is 8 (assuming f(0) = 0)", 8, '6th Fibonacci number'),
+    '10th number in the fibonacci sequence?' => build_test("The 10th fibonacci number is 55 (assuming f(0) = 0)", 55, '10th Fibonacci number'),
+    'what is the 18th number in the fibonacci series' => build_test("The 18th fibonacci number is 2584 (assuming f(0) = 0)", 2584, '18th Fibonacci number'),
+    'What is the 2nd fibonacci number' => build_test("The 2nd fibonacci number is 1 (assuming f(0) = 0)", 1, '2nd Fibonacci number'),
+    'what is the 250001 fib' => undef,
+    'is 14 a fibonacci number' => build_test("14 is not a Fibonacci number", "No", '14 is not a Fibonacci number'),
+    'is 13 in the fibonacci sequence?' => build_test("13 is a Fibonacci number", "Yes", '13 is a Fibonacci number'),
+    'is 6 a fib?' => build_test("6 is not a Fibonacci number", "No", "6 is not a Fibonacci number"),
```
Comment (GuiltyDolphin) [inline t/Fibonacci.t]: We should probably support negatives (`is -5 a fibonacci number?`).

**Comment by GuiltyDolphin** [comment]
@moollaza @shellhead This hangs with no response on Beta and production with `92nd fibonacci number` (or higher); anyone else getting this?

**Comment by mintsoft** [comment]
@GuiltyDolphin Confirmed

**Comment by GuiltyDolphin** [comment]
@mintsoft Any ideas what's going on? It's clearly nothing to do with the package seeing as it does it already on production... And the numbers aren't very big at this point.

**Comment by edgesince84** [comment]
@shellhead thanks, I'm sure we can find something for you to help us with!

**Comment by shellhead** [comment]
Is there a way I can help debug this? I ran `duckpan query` in testing, with large numbers (10^5), and didn't have issues. That sounds different than the beta environment, though.

File: t/Fibonacci.t (hunk 6 above)
Comment (koosha--) [inline t/Fibonacci.t]: @GuiltyDolphin If somebody asks whether a negative number is a Fibonacci one, I would throw an insulting message to them! Checking for negative numbers is really meaningless.

**Comment by GuiltyDolphin** [comment]
@shellhead Yeah - I think it's some internal configuration; I don't think we have much hope finding out, it'll need to be checked out internally.

/cc @zekiel @moollaza @jdorweiler

**Comment by moollaza** [comment]
@shellhead @GuiltyDolphin looking at the results in DuckPAN I'm seeing that 92+ returns a Math::BigInt object instead of text for the title -- I suspect that is part of the problem?

File: lib/DDG/Goodie/Fibonacci.pm [hunk 7]
```diff
@@ -5,37 +5,60 @@ use strict;
...
-
-    my $suf = ordsuf($n);
-    my $text_answer ="The $n$suf fibonacci number is ${fib[$n]} (assuming f(0) = 0).";
-    return $text_answer, structured_answer => {
+    return $text_ans, structured_answer => {
         data => {
-            title => $fib[$n],
-            subtitle => "$n$suf Fibonacci number"  
+            title => $title,
+            subtitle => $subtitle
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Fibonacci.pm]: @moollaza Said these are Big* on larger numbers, try using `"$title"` and `"$subtitle"` instead.

Comment (shellhead) [inline lib/DDG/Goodie/Fibonacci.pm]: I changed these as suggested. When running `duckpan query`, the subtitle still looks like a number (not wrapped in quotes).

**Comment by shellhead** [comment]
Moving the comment out from outdated diff...

I changed these as suggested. When running duckpan query, the subtitle still looks like a number (not wrapped in quotes). Then again, when I run it using the previous implementation, I see the same thing.

**Comment by moollaza** [comment]
> When running duckpan query, the subtitle still looks like a number (not wrapped in quotes). Then again, when I run it using the previous implementation, I see the same thing.

@shellhead that's how it's supposed to work :) The quotes should be forcing the `$title` into a scalar (string) context though. So instead of a `Big::*` object, we're getting the string representation.

I'll deploy to Beta to confirm 👍 

----

To make it more clear, the same thing would be happening if you did `subtitle => "The title is $title";` -- you're returning a string here, and `$title`, being inside `""` converts to its scalar context

**Comment by moollaza** [comment]
@shellhead Thanks! This LGTM :+1:

**Comment by moollaza** [comment]
@shellhead we're going to need to add a CSS file here:

```css
.zci--fibonacci .c-base__title {
    word-break: break-all;
}
```

This will force long titles to wrap instead of clip the overflow: https://beta.duckduckgo.com/?q=500th+fibonacci+number&ia=answer

**Pull request merged**